sdks/python/apache_beam/transforms/userstate.py

```diff
@@ -161,3 +176,126 @@ def validate_stateful_dofn(dofn):
             ('The on_timer callback for %s is not the specified .%s method '
              'for DoFn %r (perhaps it was overwritten?).') % (
                  timer_spec, method_name, dofn))
+
+
+class RuntimeTimer(object):
+  """Timer interface object passed to user code."""
+
+  def __init__(self, timer_spec):
+    self._cleared = False
+    self._new_timestamp = None
+
+  def clear(self):
+    self._cleared = True
+    self._new_timestamp = None
+
+  def set(self, timestamp):
+    self._new_timestamp = timestamp
+
+
+class RuntimeState(object):
+  """State interface object passed to user code."""
+
+  def __init__(self, state_spec, state_tag, current_value_accessor):
+    self._state_spec = state_spec
+    self._state_tag = state_tag
+    self._current_value_accessor = current_value_accessor
+
+  @staticmethod
+  def for_spec(state_spec, state_tag, current_value_accessor):
+    if isinstance(state_spec, BagStateSpec):
+      return BagRuntimeState(state_spec, state_tag, current_value_accessor)
+    elif isinstance(state_spec, CombiningValueStateSpec):
+      return CombiningValueRuntimeState(state_spec, state_tag,
+                                        current_value_accessor)
+    else:
+      raise ValueError('Invalid state spec: %s' % state_spec)
+
+  def _encode(self, value):
+    return self._state_spec.coder.encode(value)
+
+  def _decode(self, value):
+    return self._state_spec.coder.decode(value)
+
+  def prefetch(self):
+    # The default implementation here does nothing.
+    pass
+
+
+# Sentinel designating an unread value.
+UNREAD_VALUE = object()
+
+
+class BagRuntimeState(RuntimeState):
+  """Bag state interface object passed to user code."""
+
+  def __init__(self, state_spec, state_tag, current_value_accessor):
+    super(BagRuntimeState, self).__init__(
+        state_spec, state_tag, current_value_accessor)
+    self._cached_value = UNREAD_VALUE
+    self._cleared = False
+    self._new_values = []
+
+  def read(self):
+    if self._cached_value is UNREAD_VALUE:
+      self._cached_value = self._current_value_accessor()
+    if not self._cleared:
+      encoded_values = itertools.chain(self._cached_value, self._new_values)
+    else:
+      encoded_values = self._new_values
+    return (self._decode(v) for v in encoded_values)
+
+  def add(self, value):
+    self._new_values.append(self._encode(value))
+
+  def clear(self):
+    self._cleared = True
+    self._new_values = []
+
+
+class CombiningValueRuntimeState(RuntimeState):
```

Comment:
I wonder if this could be built as a user-level state on top of BagState, rather than having it be a sibling primitive that every runner will have to understand.

Reply:
I think that is a good idea, to reduce the number of primitives in the FnAPI / portability logic (CC: @lukecwik).  Do you remember why in Dataflow combining was its own type of state?--was there some combining happening at the shuffle level?  In some sense though, this is the "user" portion of the code, in that any state the FnAPI exposes could be used here, and other runners need only implement the BagState tag type.  It's not clear to me how we can cleanly make this more "user-level"--you can imagine having the user, instead of declaring a combining value state, declaring a bag state instead, and when they receive the runtime state object, wrapping it with something that shims it to look like a combining value state.  But this is much more confusing so I don't see it as preferable to the current approach of having this be an explicit state spec type and making the wrapping of bag state happen at this level.

Reply:
I'm thinking that the user would declare the state as CombiningState, but that simply delegates to a BagState. (More complicated states could be built up as well. Perhaps the base could be MultimapState.) Exactly what the API would look like I haven't had time to think about yet...